File: Lesson/Lesson 29 Max7219 drive dot matrix experiment/Python/Max7219.py
# -*- coding:utf-8 -*-

import RPi.GPIO as GPIO
import time

DecodeMode = 0x09 #译码模式寄存器
Intensity = 0x0a #亮度寄存器
ScanLimit = 0x0b #扫描位数寄存器
ShutDown = 0x0c #低功耗模式寄存器
DisplayTest = 0x0f #显示测试寄存器
ShutdownMode = 0x00 #低功耗方式
NormalOperation = 0x01 #正常操作方式
ScanDigit = 0x07 #扫描位数设置, 显示8位数码管
DecodeDigit = 0x00 #译码设置, 8位均为非译码
IntensityGrade = 0x0a #亮度级别设置
TestMode = 0x01 #显示测试模式
TextEnd = 0x00 #显示测试结束, 恢复正常工作模式
buffer =[ 0x78, 0xFC, 0xFE, 0x7F, 0x7F, 0xFE, 0xFC, 0x78 ]

class Max7219:

	def __init__(self, CLK, DIN, CS):
		self.DIN = DIN
		self.CS = CS
		self.CLK = CLK
		GPIO.setmode(GPIO.BCM)
		GPIO.setup(self.DIN, GPIO.OUT)
		GPIO.setup(self.CS, GPIO.OUT)
		GPIO.setup(self.CLK, GPIO.OUT)


	def start(self):
		self.writeWord(DecodeMode, 0x00)
		self.writeWord(Intensity, 0x08)
		self.writeWord(ScanLimit, 0x07)
		self.writeWord(ShutDown, 0x01)
		self.writeWord(DisplayTest, 0x00)


	def senduchar(self, ch):
		for i in range(0, 8):
			tmp = ch & 0x80
			if tmp:
				GPIO.output(self.DIN, GPIO.HIGH)
			else:
				GPIO.output(self.DIN, GPIO.LOW)
			ch = ch << 1
			GPIO.output(self.CLK, GPIO.HIGH)
			GPIO.output(self.CLK, GPIO.LOW)


	def writeWord(self, addr, num):
		GPIO.output(self.CS, GPIO.HIGH)
		GPIO.output(self.CS, GPIO.LOW)
		GPIO.output(self.CLK, GPIO.LOW)
		self.senduchar(addr)
		self.senduchar(num)
		GPIO.output(self.CS, GPIO.HIGH)

	def write(self, buff):
		for i in range(0, 8):
			self.writeWord(i + 1, buff[i])
			time.sleep(0.02)


max7219 = Max7219(22, 23, 24)
max7219.start()
while True:
	max7219.write(buffer)

File: Lesson/Lesson 29 Max7219 drive dot matrix experiment/Java/Max7219.java
import com.pi4j.wiringpi.Gpio;
import com.pi4j.wiringpi.GpioInterrupt;
import com.pi4j.wiringpi.GpioInterruptListener;
import com.pi4j.wiringpi.GpioInterruptEvent;
import com.pi4j.wiringpi.GpioUtil;


public class Max7219 {
	static int DecodeMode = 0x09; //译码模式寄存器
	static int Intensity = 0x0a; //亮度寄存器
	static int ScanLimit = 0x0b; //扫描位数寄存器
	static int ShutDown = 0x0c; //低功耗模式寄存器
	static int DisplayTest = 0x0f; //显示测试寄存器
	static int ShutdownMode = 0x00; //低功耗方式
	static int NormalOperation = 0x01; //正常操作方式
	static int ScanDigit = 0x07; //扫描位数设置, 显示8位数码管
	static int DecodeDigit = 0x00; //译码设置, 8位均为非译码
	static int IntensityGrade = 0x0a; //亮度级别设置
	static int TestMode = 0x01; //显示测试模式
	static int TextEnd = 0x00; //显示测试结束, 恢复正常工作模式
	static int DIN = 4, CS = 5, CLK = 3;
	static int buffer[] =
							{
								0x78,0xFC,0xFE,0x7F,0x7F,0xFE,0xFC,0x78,
							};

	public static void init(){
		Gpio.pinMode(DIN, Gpio.OUTPUT);
		Gpio.pinMode(CS, Gpio.OUTPUT);
		Gpio.pinMode(CLK, Gpio.OUTPUT);
		writeWord(DecodeMode, 0x00);
		writeWord(Intensity, 0x08);
		writeWord(ScanLimit, 0x07);
		writeWord(ShutDown, 0x01);
		writeWord(DisplayTest, 0x00);
	}

	public static void send_data(int ch){
		int i, tmp;
		for(i = 0; i < 8; i++){
			tmp = ch & 0x80;
			if(tmp == 0x80)
				Gpio.digitalWrite(DIN, Gpio.HIGH);
			else
				Gpio.digitalWrite(DIN, Gpio.LOW);
			ch = ch << 1;
			Gpio.digitalWrite(CLK, Gpio.HIGH);
			Gpio.digitalWrite(CLK, Gpio.LOW);
		}
	}

	public static void writeWord(int addr, int num){
		Gpio.digitalWrite(CS, Gpio.HIGH);
		Gpio.digitalWrite(CS, Gpio.LOW);
		Gpio.digitalWrite(CLK, Gpio.LOW);
		send_data(addr);
		send_data(num);
		Gpio.digitalWrite(CS, Gpio.HIGH);
	}

	public static void write(int [] buff){
		int i;
		for(i = 0; i < 8; i++){
			writeWord(i + 1, buff[i]);
			System.out.println(buff[i]);
			Gpio.delay(20);
		}
	}

	public static void main(String[] args) throws InterruptedException {
		// setup wiring pi
		if (Gpio.wiringPiSetup() == -1) {
			System.out.println(" ==>> GPIO SETUP FAILED");
			return;
		}
		init();
		for (;;) {
			write(buffer);
		}
	}
}
